src/librustc_typeck/collect.rs

```diff
@@ -832,16 +835,22 @@ fn convert_variant(
             let fid = tcx.hir().local_def_id(f.hir_id);
             let dup_span = seen_fields.get(&f.ident.normalize_to_macros_2_0()).cloned();
             if let Some(prev_span) = dup_span {
-                struct_span_err!(
-                    tcx.sess,
-                    f.span,
-                    E0124,
-                    "field `{}` is already declared",
-                    f.ident
-                )
-                .span_label(f.span, "field already declared")
-                .span_label(prev_span, format!("`{}` first declared here", f.ident))
-                .emit();
+                #[derive(SessionDiagnostic)]
+                #[code = "E0124"]
+                struct FieldAlreadyDeclared {
+                    field_name: String,
+                    #[error = "field `{field_name}` is already declared"]
+                    #[label = "field already declared"]
+                    span: Span,
+                    #[label = "`{field_name}` first declared here"]
+                    prev_span: Span,
+                }
+
+                tcx.sess.emit_err(FieldAlreadyDeclared {
+                    field_name: f.ident.to_string(),
+                    span: f.span,
+                    prev_span,
+                });
```

Comment:
Here's an existing diagnostic I've replaced, which demonstrates an actual usage of the derive macro.